daemon/containerd/image_delete.go

```diff
@@ -205,19 +239,61 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
 	})
 }
 
-// isSingleReference returns true if there are no other images in the
+// getSameReferences returns true if there are no other images in the
 // daemon targeting the same content as `img` that are not dangling.
-func (i *ImageService) isSingleReference(ctx context.Context, img images.Image) (bool, error) {
-	refs, err := i.client.ImageService().List(ctx, "target.digest=="+img.Target.Digest.String())
-	if err != nil {
-		return false, err
+func (i *ImageService) getSameReferences(ctx context.Context, img *images.Image, imgs []images.Image) ([]images.Image, error) {
+	var (
+		named      reference.Named
+		tag        string
+		sameRef    []images.Image
+		digestRefs = []images.Image{}
+	)
+	if img != nil {
+		repoRef, err := reference.ParseNamed(img.Name)
+		if err != nil {
+			return nil, err
+		}
+		named = repoRef
+		if tagged, ok := named.(reference.Tagged); ok {
+			tag = tagged.Tag()
+		}
 	}
-	for _, ref := range refs {
-		if !isDanglingImage(ref) && ref.Name != img.Name {
-			return false, nil
+	for _, ref := range imgs {
+		if !isDanglingImage(ref) {
+			if repoRef, err := reference.ParseNamed(ref.Name); err == nil {
+				if named == nil {
+					named = repoRef
+					if tagged, ok := named.(reference.Tagged); ok {
+						tag = tagged.Tag()
+					}
+				} else if named.Name() != repoRef.Name() {
+					continue
+				} else if tagged, ok := repoRef.(reference.Tagged); ok {
+					if tag == "" {
+						tag = tagged.Tag()
+					} else if tag != tagged.Tag() {
+						// Same repo, different tag, do not include digest refs
+						digestRefs = nil
+						continue
+					}
+				} else {
+					if digestRefs != nil {
+						digestRefs = append(digestRefs, ref)
+					}
+					// Add digest refs at end if no other tags in the same name
+					continue
+				}
+			} else {
+				// Ignore names which do not parse
+				log.G(ctx).WithError(err).WithField("image", ref.Name).Info("failed to parse image name, ignoring")
```

Comment:
Use `WithFields`?

Reply:
I don't usually use `WithFields` for errors and the error key is not aliased in the `log` package. There might have been a place I used `WithField` twice and should update to `WithFields` though.